astropy/units/physical.py

```diff
@@ -1,133 +1,472 @@
+            return self._physical_type_id == other._physical_type_id
+        elif isinstance(other, str):
+            other = _standardize_physical_type_names(other)
+            return other.issubset(self._physical_type)
+        else:
+            return NotImplemented
+
+    def __ne__(self, other):
+        equality = self.__eq__(other)
+        return not equality if isinstance(equality, bool) else NotImplemented
+
+    @property
+    def _name_string_as_ordered_set(self):
+        return "{" + str(self._physical_type_list)[1:-1] + "}"
+
+    def __repr__(self):
+        if len(self._physical_type) == 1:
+            names = "'" + self._physical_type_list[0] + "'"
+        else:
+            names = self._name_string_as_ordered_set
+        return f"PhysicalType({names})"
+
+    def __str__(self):
+        if len(self._physical_type) == 1:
+            return self._physical_type_list[0]
+        else:
+            return self._name_string_as_ordered_set
+
+    @staticmethod
+    def _dimensionally_compatible_unit(obj):
+        """
+        If a unit is passed in, return that unit.  If a physical type is
+        passed in, return a unit that corresponds to that physical type.
+        If a real number is passed in, return a dimensionless unit.
+        Otherwise, return `NotImplemented`.
+        """
+        if isinstance(obj, core.UnitBase):
+            return _replace_temperatures_with_kelvin(obj)
+        elif isinstance(obj, PhysicalType):
+            return obj._unit
+        elif isinstance(obj, numbers.Real):
+            return core.dimensionless_unscaled
+        else:
+            return NotImplemented
+
+    def _dimensional_analysis(self, other, operation):
+        other_unit = self._dimensionally_compatible_unit(other)
+        if other_unit is NotImplemented:
+            return NotImplemented
+        other_unit = _replace_temperatures_with_kelvin(other_unit)
+        new_unit = getattr(self._unit, operation)(other_unit)
+        return new_unit.physical_type
+
+    def __mul__(self, other):
+        return self._dimensional_analysis(other, "__mul__")
+
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+    def __truediv__(self, other):
+        return self._dimensional_analysis(other, "__truediv__")
+
+    def __rtruediv__(self, other):
+        other = self._dimensionally_compatible_unit(other)
+        if other is NotImplemented:
+            return NotImplemented
+        return other.physical_type._dimensional_analysis(self, "__truediv__")
+
+    def __pow__(self, power):
+        return (self._unit ** power).physical_type
+
+    def __hash__(self):
+        return hash(self._physical_type_id)
+
+    def __len__(self):
+        return len(self._physical_type)
+
+    def __array_ufunc__(self):
+        # This prevents operations like Unit(...) * PhysicalType(...)
+        # from returning a Quantity with a PhysicalType instance as the
+        # Quantity's value.
+        pass
```

Comment:
My motivation for adding this method was to try ways to prevent things like this from happening:
```Python
>>> u.m * u.m.physical_type
<Quantity PhysicalType('length') m>
```
I wanted `Unit.__mul__` to return `NotImplemented` here so that this operation would fall back on `PhysicalType.__rmul__`, ideally without making any changes to `Unit` or `Quantity`.  I speculated that having `PhysicalType.__array_ufunc__` raise a `TypeError` would prevent this from happening, but it appears to only be necessary for `PhysicalType.__array_ufunc__` to exist without actually doing anything.  It's a bit confusing why this works, so I'm curious if there is a more understandable way to ensure this behavior.

Reply:
I'll think about this - I think the underlying problem is `np.array(physical_type)` creating an object array. `__array__ = None` may be better.

Reply:
Thanks!  I'll go with that for the time being.